src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
-
-    int* d_td_d_offd_i = new int[ncdofs + 1];
-    int d_td_d_offd_nnz = 0;
-    for (int i = 0; i < ncdofs; i++)
+std::vector<DenseMatrix> GraphCoarsen::BuildElemM(const MixedMatrix& mgl,
+                                                  const SparseMatrix& agg_cdof_edge) const
+{
+    int num_aggs = gt_.agg_ext_edge_.Rows();
+    int num_faces = gt_.face_edge_local_.Rows();
+
+    std::vector<DenseMatrix> M_el(num_aggs);
+
+    DenseMatrix h_F_potential;
+    DenseMatrix h_D_trace;
+    DenseMatrix bub_block;
+    DenseMatrix ortho_vects;
+    DenseMatrix DTT_F;
+    DenseMatrix DTT_B;
+
+    for (int agg = 0; agg < num_aggs; ++agg)
     {
-        d_td_d_offd_i[i] = d_td_d_offd_nnz;
-        if (d_td_d_tmp_offd.RowSize(i))
-            d_td_d_offd_nnz++;
-    }
-    d_td_d_offd_i[ncdofs] = d_td_d_offd_nnz;
-    int* d_td_d_offd_j = new int[d_td_d_offd_nnz];
-    d_td_d_offd_nnz = 0;
-
-    mfem::SparseMatrix face_d_td_d_offd;
-    HYPRE_Int* junk_map;
-    face_d_td_d.GetOffd(face_d_td_d_offd, junk_map);
-
-    int face_1st_cdof, face_ncdofs;
-    int* face_cdof_i = face_cdof.GetI();
-    int* face_cdof_j = face_cdof.GetJ();
-    mfem::Array<int> face_cdofs;
-    for (int i = 0; i < nfaces; i++)
-        if (face_d_td_d_offd.RowSize(i))
+        linalgcpp::HStack(F_potential_[agg], h_F_potential);
+        linalgcpp::HStack(D_trace_[agg], h_D_trace);
+
+        int num_bubbles = vertex_targets_[agg].Cols() - 1;
+        int num_traces = h_F_potential.Cols();
+
+        M_el[agg].SetSize(num_bubbles + num_traces);
+        M_el[agg] = 0.0;
+
+        if (num_bubbles > 0)
         {
-            face_ncdofs = face_cdof_i[i + 1] - face_cdof_i[i];
-            face_1st_cdof = face_cdof_j[face_cdof_i[i]];
-            GetTableRow(d_td_d_tmp_offd, face_1st_cdof, face_cdofs);
-            assert(face_cdofs.Size() == face_ncdofs);
-            for (int j = 0; j < face_ncdofs; j++)
-                d_td_d_offd_j[d_td_d_offd_nnz++] = face_cdofs[j];
+            int total_cols = vertex_targets_[agg].Cols();
+            vertex_targets_[agg].GetCol(1, total_cols, ortho_vects);
+
+            ortho_vects.MultAT(B_potential_[agg], bub_block);
+            M_el[agg].SetSubMatrix(0, 0, bub_block);
         }
-    assert(d_td_d_offd_i[ncdofs] == d_td_d_offd_nnz);
-    mfem::SparseMatrix d_td_d_offd(d_td_d_offd_i, d_td_d_offd_j,
-                                   d_td_d_diag_data.GetData(), ncdofs, d_td_d_offd_nnz,
-                                   true, false, false);
-
-    mfem::HypreParMatrix d_td_d(
-        comm, edge_cd_start_.Last(), edge_cd_start_.Last(), edge_cd_start_,
-        edge_cd_start_, &d_td_d_diag, &d_td_d_offd, d_td_d_map);
-
-    // Create a selection matrix to set dofs on true faces to be true dofs
-    int* select_i = new int[ncdofs + 1];
-    select_i[0] = 0;
-    int cdof_counter = 0;
-    for (int i = 0; i < nfaces; i++)
+
+        h_D_trace.MultAT(h_F_potential, DTT_F);
+        h_D_trace.MultAT(B_potential_[agg], DTT_B);
```

Comment:
Can we implement (potential * D) * trace instead of potential * (D * trace), that is, computing (potential * D) first. The reason is because (I think) the former has higher chance to be faster. This depends on whether potential or trace is longer in length (which in turn depends on the graph and partition). At least for graphs coming from conforming meshes, I think trace is always shorter. 

This is the reason why the original code computes (potential * D) first. That said, I think the timing difference will be outweighed by the solving time of local eigen problems, so I am not too worried about it, but I just think if possible, why not. I guess the drawback would be, the code will be not as clean as the current version. The main problem is traces are not of the same length, so here will need an extra loop over faces. I just want to bring this up and you can think about it.

Reply:
Yes, this may be faster.  Right now the ML code uses local RAP anyways, so I will have to revisit this idea when reimplementing assembling local `M` w/ potentials.